app/controllers/blogs_controller.rb

```diff
@@ -3,13 +3,16 @@
 class BlogsController < ApplicationController
   skip_before_action :authenticate_user!, only: %i[index show]
 
-  before_action :set_blog, only: %i[show edit update destroy]
+  before_action :set_blog, only: %i[edit update destroy]
 
   def index
     @blogs = Blog.search(params[:term]).published.default_order
   end
 
-  def show; end
+  def show
+    @blog = Blog.find(params[:id])
+    @blog = Blog.find(nil) if @blog.secret && current_user != @blog.user
```

Comment:
```suggestion
    @blog = Blog.hogehoge.find(params[:id])
```

scopeを使ってうまくSQL一発で取得できるようにするとベストです。

Reply:
blogモデルにvisible_userスコープを追加しました。
```ruby
scope :visible_user, lambda { |user|
    if user.present?
      where('secret = ? OR user_id = ?', false, user.id)
    else
      where(secret: false)
    end
  }
  ```